url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/ui/VirtualTerminal.java

```diff
@@ -0,0 +1,386 @@
+    /**
+     * ストリームを閉じ、リソースを解放します。
+     */
+    public void close() {
+        try {
+            setHalt(true);
+            inPin.close();
+            inPout.close();
+            outPin.close();
+            outPout.close();
+        } catch (IOException e) {
+            //ignore
+        }
+    }
+
+    /**
+     * 表示を開始する行数を取得します。
+     *
+     * @return 表示を開始する行数
+     */
+    public int getStartLine() {
+        return startLine;
+    }
+
+    /**
+     * 表示を開始する行数を設定します。
+     *
+     * @param l 表示を開始する行数
+     */
+    public void setStartLine(int l) {
+        startLine = l;
+    }
+
+    @Override
+    public void mouseClicked(MouseEvent e) {
+        switch (e.getButton()) {
+        case MouseEvent.BUTTON3:
+            //右クリックでクリップボードから貼り付け
+            Clipboard clip = getToolkit().getSystemClipboard();
+            String strClip = getClipboardText(clip);
+
+            if (!strClip.equals("")) {
+                try {
+                    inPout.write(strClip.getBytes());
+                } catch (IOException ie) {
+                    //do nothing
+                }
+            }
+            break;
+        }
+    }
+
+    private String getClipboardText(Clipboard clip) {
+        Transferable trans = clip.getContents(null);
+        String strClip;
+
+        if (trans == null || !trans.isDataFlavorSupported(DataFlavor.stringFlavor)) {
+            strClip = "";
+        } else {
+            try {
+                strClip = (String)trans.getTransferData(DataFlavor.stringFlavor);
+            } catch (UnsupportedFlavorException | IOException ue) {
+                strClip = "";
+            }
+        }
+
+        return strClip;
+    }
+
+    @Override
+    public void mousePressed(MouseEvent e) {
+        //do nothing
+    }
+
+    @Override
+    public void mouseReleased(MouseEvent e) {
+        //do nothing
+    }
+
+    @Override
+    public void mouseEntered(MouseEvent e) {
+        //do nothing
+    }
+
+    @Override
+    public void mouseExited(MouseEvent e) {
+        //do nothing
+    }
+
+    @Override
+    public void mouseDragged(MouseEvent e) {
+        /*
+        e.translatePoint(-vt.getBoxScreen().getContents().x, -vt.getBoxScreen().getContents().y);
+        int x = e.getX() / vt.getBoxChar().getWidth();
+        int y = e.getY() / vt.getBoxChar().getHeight();
+        y += getStartLine();
+        x = Math.max(x, 0);
+        x = Math.min(x, vt.getColumns() - 1);
+        y = Math.max(y, 0);
+        y = Math.min(y, vt.getMaxLines() - 1);
+        System.out.printf("char (%d, %d) %c\n", x, y, vt.getChar(x, y));
+        */
+    }
+
+    @Override
+    public void mouseMoved(MouseEvent e) {
+        //do nothing
+    }
+
+    @Override
+    public void mouseWheelMoved(MouseWheelEvent e) {
+        scr.setValue(scr.getValue() + e.getWheelRotation() * 3);
+    }
+
+    @Override
+    public void keyTyped(KeyEvent e) {
+        //do nothing
+    }
+
+    @Override
+    public void keyPressed(KeyEvent e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 249-262 into a separate method.